### PR TITLE
Include SWI Prolog binary in package

### PR DESCRIPTION
**Pull request by lucksus**
[Depends on https://github.com/perspect3vism/ad4m-executor/pull/83, hence v0.1.41-alpha-X dependency]

When ad4m-executor is packaged into a vercel/pkg in ad4m-host, Prolog queries don't work because swipl isn't installed and in $PATH in that context. These changes make sure that swipl is set up and used correctly in ad4m-host without requiring the user to install it manually on their machine.

## High-level overview of needed steps
1. Include the whole swipl installation in vercel package (copy over to `temp`)
2. Copy swipl to .ad4m directory on executor init
3. Provide swipl paths to executor

## What's also needed to make it work
* Change .so/.dylib paths in swipl binary to have dependencies be found through relative links in ~/.ad4m/swipl (different for Linux vs. macOS, not needed on Windows since .dll are already next to swipl.exe). Done in `scripts/relink-swipl-<os>.sh`.
* Copy node-swipl-stdio's `top.pl` to swipl homedir in ~/.ad4m/swipl as well since swipl binary can't find that through vercel's /snapshot filesystem (see changes to our node-swipl-stdio fork: https://github.com/perspect3vism/node-swipl-stdio/compare/pre-ad4m-host...master)
* Change ad4m config directory path from %APP_DATA% to user's home directory since swipl can't deal with whitespaces in paths to its homedir (-> problem on macOS with `~/Library/Application Support/ad4m` which now is `~/.ad4m/`). See https://github.com/perspect3vism/ad4min/pull/28.

## Progress

- [x] script for copying from nix (Linux/macOS)
- [x] script for windows
- [x] add SWI license text (https://www.swi-prolog.org/license.html)
- [x] add steps to GH CI
- [x]  change ad4m-executor dep version after merging https://github.com/perspect3vism/ad4m-executor/pull/83 and making new (non-alpha) release.

### CODE REVIEW

**Review by kaichaosun — APPROVED** [review]
Generally looks good to me!
Two thoughts though, 
1. @fayeed mentioned that ad4m-test library use ad4m-host and create different folders for test data, if using `~/.ad4m` may broken ad4m-test?
2. I have a feeling that using binaries (holochain, swipl) is not a perfect solution to build desktop apps like ad4min, that's why I'm thinking rebuild ad4m in Rust is ideal, since use holochain as a Rust library should be feasible (not confirmed yet though). I just found this [prolog implementation](https://github.com/mthom/scryer-prolog) in Rust is interesting, do you evaluate it to be used in ad4m?

File: src/commands/init.ts [hunk 1]
```diff
@@ -4,16 +4,17 @@
...
-import { getAppDataPath } from "appdata-path";
 import utils from 'util';
 import { fetchLatestBootstrapSeed, MAINNET_SEED } from '../utils/fetchLatestBootstrapSeed';
 import { CONFIG, getConfig } from '../utils/config';
 import ReadlineSync from 'readline-sync';
 import os from 'os'
+import { ad4mDataDirectory } from '../ad4mDataDirectory';
 
 const copyFile = utils.promisify(fs.copyFile);
+const copyDir = utils.promisify(fs.copy)
```
Comment (fayeed) [inline src/commands/init.ts:17]: this fs functions already return a promise

File: src/ad4mDataDirectory.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { homedir } from 'os'
+import path from 'path'
+
+export function ad4mDataDirectory(override = undefined) {
```
Comment (fayeed) [inline src/ad4mDataDirectory.ts]:
```suggestion
export function ad4mDataDirectory(override?: string) {
```

**Review by fayeed — APPROVED** [review]
Left a few comments but looks good.

**Comment by fayeed** [comment]
One more thing ad4m-test will need to be updated to use the home directory, right now it uses data path.

**Comment by fayeed** [comment]
> 1. @fayeed mentioned that ad4m-test library use ad4m-host and create different folders for test data, if using `~/.ad4m` may broken ad4m-test?

Yeah, ad4m-test will also have to be updated to use the home directory but I see that there is a way to override the `.ad4m` path in the ad4m-host so shouldn't be a problem.

**Comment by lucksus** [comment]
Thanks for your reviews.

@kaichaosun,
1. Concerning ad4m-test: yes, ad4m-host understands a `dataPath` config but AFAIK that only changes the name of the data directory (`ad4m` vs. `flux`) but not where that is put. I was fixed to the XDG app data path and now is to HOMEDIR. We might want to change ad4m-host to allow for complete overriding (if that is not the case yet).
2. Fully agree. I did look at scryer when searching for a Prolog engine and if we had a Rust implementation of AD4M I'd prefer using that (and Holochain) compiled into our executable, yes. Then I found [this WASM build](https://github.com/SWI-Prolog/swipl-wasm) of SWI-Prolog and tried integrating that which first seemed to work but then let to problems I couldn't really (easily) debug, so I switched to the swipl binary and node-swipl-stdio. I don't think we would get much from switching the swipl binary to scryer binary (as swipl seems much more mature), but once we have rebuild everything in Rust (which I really think is a great idea, but not something I would like to put on our plate in the immediate future), I would love take a look at Scryer again.

**Pull request merged**